arangod/Aql/IResearchViewExecutor.tpp

```diff
@@ -1412,122 +1439,265 @@ bool IResearchViewHeapSortExecutor<ExecutionTraits>::fillBufferInternal(
+  this->_indexReadBuffer.preAllocateStoredValuesBuffer(
+      atMost, this->_infos.getScoreRegisters().size(),
+      this->_infos.getOutNonMaterializedViewRegs().size());
+  this->_indexReadBuffer.setForParallelAccess(
+      atMost, this->_infos.getScoreRegisters().size(),
+      this->_infos.getOutNonMaterializedViewRegs().size());
+  results.reserve(parallelism - 1);
+  // we must wait for our threads before bailing out
+  // as we most likely will release segments and
+  // "this" will also be invalid.
+  auto cleanupThreads = irs::Finally([&]() noexcept {
+    results.erase(std::remove_if(results.begin(), results.end(),
+                                 [](auto& f) { return !f.valid(); }),
+                  results.end());
+    if (results.empty()) {
+      return;
+    }
+    auto runners = futures::collectAll(results);
+    runners.wait();
+  });
+  while (bufferIdx.load() < atMostInitial) {
+    TRI_ASSERT(results.empty());
+    size_t i = 0;
+    size_t selfExecute{std::numeric_limits<size_t>::max()};
+    auto toFetch = atMost - bufferIdx.load();
+    while (toFetch && results.size() < (parallelism - 1) &&
+           i < _segmentReaders.size()) {
+      auto& reader = _segmentReaders[i];
+      if (!reader.itr) {
+        if (_segmentOffset >= count) {
+          // no new segments. But maybe some existing readers still alive
+          ++i;
+          continue;
+        }
+        reader.readerOffset = _segmentOffset++;
+      }
+      reader.atMost = std::min(atMostInitial, toFetch);
+      toFetch -= reader.atMost;
+      if (selfExecute < parallelism) {
+        futures::Promise<bool> promise;
+        auto future = promise.getFuture();
+        if (ADB_UNLIKELY(!readersPool.run(
+                [&, ctx = &reader, pr = std::move(promise)]() mutable {
+                  try {
+                    pr.setValue(readSegment<true>(*ctx, bufferIdx));
+                  } catch (...) {
+                    pr.setException(std::current_exception());
+                  }
+                }))) {
+          TRI_ASSERT(false);
+          THROW_ARANGO_EXCEPTION_MESSAGE(
+              TRI_ERROR_INTERNAL,
+              " Failed to schedule parallel search view reading");
+        }
+        // this should be noexcept as we've reserved above
+        results.push_back(std::move(future));
+      } else {
+        selfExecute = i;
+      }
+      ++i;
+    }
+    if (selfExecute < std::numeric_limits<size_t>::max()) {
+      gotData |= readSegment<true>(_segmentReaders[selfExecute], bufferIdx);
+    } else {
+      TRI_ASSERT(results.empty());
       break;
     }
+    // we run this in noexcept mode as with current implementation
+    // we can not recover and properly wait for finish in case
+    // of exception in the middle of collectAll or wait.
+    [&]() noexcept {
+      auto runners = futures::collectAll(results);
+      runners.wait();
+      for (auto& r : runners.result().get()) {
+        gotData |= r.get();
+      }
+    }();
```

Comment:
You can use WaitGroup to avoid allocation here.

It exists in iresearch IndexWriter

Reply:
In current state WaitGroup is private inside IndexWriter. As we anyway planning a follow up PR  with iresearch submodule update - let's plan that switch on that PR